sql/src/main/java/io/crate/operation/collect/files/CSVLineParser.java

```diff
@@ -36,51 +38,35 @@
 
 public class CSVLineParser {
 
-    private static CsvMapper csvMapper = new CsvMapper();
     private static List<Object> keyList;
     private static ObjectMapper objectMapper = new ObjectMapper();
+    private static CsvMapper csvMapper = new CsvMapper();
+    private static CsvObjectReader csvReader = csvMapper.enable(CsvParser.Feature.TRIM_SPACES)
+        .readerWithTypedSchemaFor(String.class);
 
-    public void parseHeader(String header) throws IOException {
-        keyList = csvMapper
-            .enable(CsvParser.Feature.TRIM_SPACES)
-            .readerWithSchemaFor(String.class)
-            .readValues(header.getBytes(StandardCharsets.UTF_8))
-            .readAll();
-
-        Set<Object> keySet  = new HashSet<>(keyList);
+    public void parseHeader(BufferedReader currentReader) throws IOException {
+        String header = currentReader.readLine();
+        keyList = csvReader.readValues(header.getBytes(StandardCharsets.UTF_8)).readAll();
+        Set<Object> keySet = new HashSet<>(keyList);
         if (keySet.contains("")) {
             keySet.removeAll(Collections.singleton(""));
         }
 
-        if (keySet.size() != keyList.size()) {
+        if (keySet.size() != keyList.size() || keySet.size() == 0) {
             throw new IllegalArgumentException("Invalid keys!");
         }
     }
 
     public byte[] parse(String row) throws IOException {
-        List<Object> recordList = csvMapper
-            .enable(CsvParser.Feature.TRIM_SPACES)
-            .readerWithTypedSchemaFor(String.class)
-            .readValues(row.getBytes(StandardCharsets.UTF_8))
-            .readAll();
-
-        return convertCSVToMap(recordList);
-    }
-
-    private byte[] convertCSVToMap(List<Object> recordList) throws JsonProcessingException {
-
-        if (recordList.isEmpty()) {
-            return objectMapper.writeValueAsBytes(Collections.emptyMap());
-        }
-
-        if (keyList == null || keyList.size() != recordList.size()) {
-            throw new IllegalArgumentException("Number of row entries is not equal to the number of columns");
-        }
-
+        MappingIterator<Object> iterator = csvReader.readValues(row.getBytes(StandardCharsets.UTF_8));
         HashMap<Object, Object> csvAsMap = new HashMap<>();
```

Comment:
I think we can avoid this step by using all the features of the Jackson library:

We can create a schema when we read the header, e.g.:

```java
CsvSchema.Builder schema = new CsvSchema.Builder();
for (String value : map.keySet())
{
    schema.addColumn(value, CsvSchema.ColumnType.STRING);
}
new CsvMapper().writerFor(Map.class).with(schema.build());
```

This gives us the map directly.